tests/crc_test.c

```diff
@@ -0,0 +1,154 @@
+            .len = 1,
+            .ptr = input->ptr + i,
+        };
+        aws_hash_update(crc2, &input_i);
+    }
+    uintptr_t result2 = (uintptr_t)crc2->impl;
+    ASSERT_HEX_EQUALS(expected, result2, "one byte at a time %s(%s)", func_name, data_name);
+    aws_hash_destroy(crc);
+    aws_hash_destroy(crc1);
+    aws_hash_destroy(crc2);
+    return AWS_OP_SUCCESS;
+}
+
+/* helper function that groups crc32 tests*/
+static int s_test_known_crc32(const char *func_name, crc_fn *func, struct aws_allocator *allocator) {
+    int res = 0;
+    uint8_t *mut_DATA_32_ZEROS = (uint8_t *)DATA_32_ZEROS;
+    uint8_t *mut_DATA_32_VALUES = (uint8_t *)DATA_32_VALUES;
+    uint8_t *mut_TEST_VECTOR = (uint8_t *)TEST_VECTOR;
+    const struct aws_byte_cursor input = {
+        .len = sizeof(DATA_32_ZEROS),
+        .ptr = mut_DATA_32_ZEROS,
+    };
+    const struct aws_byte_cursor input1 = {
+        .len = sizeof(DATA_32_VALUES),
+        .ptr = mut_DATA_32_VALUES,
+    };
+    const struct aws_byte_cursor input2 = {
+        .len = sizeof(TEST_VECTOR),
+        .ptr = mut_TEST_VECTOR,
+    };
+    res |= s_test_known_crc(func_name, func, "DATA_32_ZEROS", &input, KNOWN_CRC32_32_ZEROES, allocator);
+    res |= s_test_known_crc(func_name, func, "DATA_32_VALUES", &input1, KNOWN_CRC32_32_VALUES, allocator);
+    res |= s_test_known_crc(func_name, func, "TEST_VECTOR", &input2, KNOWN_CRC32_TEST_VECTOR, allocator);
+    return res;
+}
+
+/* helper function that groups crc32c tests*/
+static int s_test_known_crc32c(const char *func_name, crc_fn *func, struct aws_allocator *allocator) {
+    int res = 0;
+    uint8_t *mut_DATA_32_ZEROS = (uint8_t *)DATA_32_ZEROS;
+    uint8_t *mut_DATA_32_VALUES = (uint8_t *)DATA_32_VALUES;
+    uint8_t *mut_TEST_VECTOR = (uint8_t *)TEST_VECTOR;
+    const struct aws_byte_cursor input = {
+        .len = sizeof(DATA_32_ZEROS),
+        .ptr = mut_DATA_32_ZEROS,
+    };
+    const struct aws_byte_cursor input1 = {
+        .len = sizeof(DATA_32_VALUES),
+        .ptr = mut_DATA_32_VALUES,
+    };
+    const struct aws_byte_cursor input2 = {
+        .len = sizeof(TEST_VECTOR),
+        .ptr = mut_TEST_VECTOR,
+    };
+    res |= s_test_known_crc(func_name, func, "DATA_32_ZEROS", &input, KNOWN_CRC32C_32_ZEROES, allocator);
+    res |= s_test_known_crc(func_name, func, "DATA_32_VALUES", &input1, KNOWN_CRC32C_32_VALUES, allocator);
+    res |= s_test_known_crc(func_name, func, "TEST_VECTOR", &input2, KNOWN_CRC32C_TEST_VECTOR, allocator);
+
+    /*this tests three things, first it tests the case where we aren't 8-byte aligned*/
+    /*second, it tests that reads aren't performed before start of buffer*/
+    /*third, it tests that writes aren't performed after the end of the buffer.*/
+    /*if any of those things happen, then the checksum will be wrong and the assertion will fail */
+    uint8_t *s_non_mem_aligned_vector;
+    s_non_mem_aligned_vector = malloc(sizeof(DATA_32_VALUES) + 6);
```

Comment:
I know it's copy/paste from aws-checksum, but still feels like we should avoid use malloc/free and use `aws_mem_acquire`/`aws_mem_release` instead, which is safer. https://github.com/awslabs/aws-c-common/blob/main/include/aws/common/allocator.h#L54-L58